qualtran/qref_interop/_bloq_to_qref.py

```diff
@@ -0,0 +1,311 @@
+    """Converts Bloq to QREF SchemaV1 object."""
+    return SchemaV1(version="v1", program=bloq_to_routine(obj))
+
+
+@singledispatch
+def bloq_to_routine(obj, name: Optional[str] = None):
+    """Import object from Qualtran by converting it into corresponding QREF RoutineV1 object.
+
+    Args:
+        obj: object to be imported. Can be either Bloq or BloqInstance.
+        name: optional name override. This can be useful e.g. if you are converting
+            CompositeBloq (which do not store meaningful names in Qualtran) and
+            know some good name for it.
+
+    Return:
+        A Bartiq object corresponding to the source Qualtran object. For both Bloqs
+        and BloqInstances the returned object is of type RoutineV1.
+
+    """
+    raise NotImplementedError(f"Cannot import object {obj} of type {type(obj)}.")
+
+
+@bloq_to_routine.register
+def _composite_bloq_to_routine(bloq: CompositeBloq, name: Optional[str] = None) -> RoutineV1:
+    """Import CompositeBloq from Qualtran.
+
+    See `import_from_qualtran` for more info.
+    """
+    return RoutineV1(
+        **_extract_common_bloq_attributes(bloq, name),
+        children=[bloq_to_routine(instance) for instance in bloq.bloq_instances],
+        connections=[_import_connection(c) for c in bloq.connections],
+    )
+
+
+@bloq_to_routine.register
+def _bloq_to_routine(bloq: Bloq, name: Optional[str] = None) -> RoutineV1:
+    """Import Bloq (other than CompositeBloq) from Qualtran.
+
+    See `import_from_qualtran` for moe info.
+    """
+    return RoutineV1(**_extract_common_bloq_attributes(bloq, name))
+
+
+@bloq_to_routine.register
+def _bloq_instance_to_routine(instance: BloqInstance) -> RoutineV1:
+    """Import Bloq (other than CompositeBloq) from Qualtran.
+
+    When importing BloqInstance we derive name from Bloq's default name and
+    instance ID to prevent duplication of names between siblings.
+
+    See `import_from_qualtran` and `_bloq_instance_name` for more info.
+    """
+    return bloq_to_routine(instance.bloq, name=_bloq_instance_name(instance))
+
+
+def _names_and_dir_from_register(reg: Register) -> Iterable[tuple[str, str]]:
+    """Yield names and directions of Bartiq Ports corresponding to QualtranRegister.
+
+    For LEFT/RIGHT registers we yield one pair of name and direction corresponding
+    of resp. output and input port. For THRU registers we yield both such pairs,
+    which effectively splits THRU registers into two ports.
+    """
+    if reg.side != Side.LEFT:
+        yield (f"out_{reg.name}", "output")
+    if reg.side != Side.RIGHT:
+        yield (f"in_{reg.name}", "input")
+
+
+def _expand_name_if_needed(reg_name, shape) -> Iterable[str]:
+    """Given a register name, expand it into sequence of names if it has nontrivial shape.
+
+    Examples:
+        "reg", () -> "reg"
+        "reg", (3,) -> "reg_0", "reg_1", "reg_2"
+    """
+    return (reg_name,) if shape == () else (f"{reg_name}_{i}" for i in range(shape[0]))
```

Comment:
Should this be expanded recursively? We can also have registers with n-dimensional shapes. For example

```
"reg", (2, 3) -> "reg_0_0", "reg_0_1", "reg_0_2", "reg_1_0", "reg_1_1", "reg_1_2", 
```

Reply:
Resolved in another discussion.